ops/framework.py

```diff
@@ -979,21 +998,35 @@ def _reemit(self, single_event_path: Optional[str] = None):
                     # and is also not a lifecycle (framework-emitted) event,
                     # it must be a custom event
                     logger.debug('Emitting custom event %s.', event)
+                else:
+                    pass
 
                 custom_handler = getattr(observer, method_name, None)
                 if custom_handler:
                     event_is_from_juju = isinstance(event, charm.HookEvent)
                     event_is_action = isinstance(event, charm.ActionEvent)
-                    with self._event_context(event_handle.kind):
-                        if (
-                            event_is_from_juju or event_is_action
-                        ) and self._juju_debug_at.intersection({'all', 'hook'}):
-                            # Present the welcome message and run under PDB.
-                            self._show_debug_code_message()
-                            pdb.runcall(custom_handler, event)
-                        else:
-                            # Regular call to the registered method.
-                            custom_handler(event)
+                    synthetic = not event_is_from_juju and not event_is_action
+                    event_module = event.__class__.__module__
+                    if event_module.startswith('ops.'):
+                        # ops.charm.Events are re-exported through ops
+                        event_module = 'ops'
+                    event_class = f'{event_module}.{event.__class__.__qualname__}'
+                    obs_class = observer_path.split('/')[-1].split('[')[0]
+                    with tracer.start_as_current_span(f'{event_handle.kind}: {obs_class}') as span:
+                        span.set_attribute('deferred', single_event_path is None)
```

Comment:
My PR to have a separate charm instance for each event will break this logic. Whoever merges second gets to fix it, I guess :)

Reply:
What would be the best way to determine if we're processing a previously deferred event?

Reply:
I think we should make it explicit. I've added an argument for `_reemit` in my PR that's set explicitly when doing either a 'natural' emit or a deferred one.